desc/grid.py

```diff
@@ -48,24 +49,59 @@ class Grid(IOAble):
         "_inverse_rho_idx",
         "_inverse_theta_idx",
         "_inverse_zeta_idx",
-        "_num_rho",
-        "_num_theta",
-        "_num_zeta",
     ]
 
     def __init__(self, nodes, sort=True):
+        # Python 3.3 (PEP 412) introduced key-sharing dictionaries.
+        # This change measurably reduces memory usage of objects that
+        # define all attributes in their __init__ method.
         self._NFP = 1
         self._sym = False
         self._node_pattern = "custom"
-
         self._nodes, self._spacing = self._create_nodes(nodes)
-
         self._enforce_symmetry()
         if sort:
             self._sort_nodes()
-        self._find_axis()
-        self._count_nodes()
-        self._scale_weights()
+        self._axis = self._find_axis()
+        (
+            self._unique_rho_idx,
+            self._inverse_rho_idx,
+            self._unique_theta_idx,
+            self._inverse_theta_idx,
+            self._unique_zeta_idx,
+            self._inverse_zeta_idx,
+        ) = self._find_unique_inverse_nodes()
+        self._L = self.num_rho
+        self._M = self.num_theta
+        self._N = self.num_zeta
+        self._weights = self._scale_weights()
+
+    def _create_nodes(self, nodes):
+        """Allow for custom node creation.
+
+        Parameters
+        ----------
+        nodes : ndarray of float, size(num_nodes,3)
+            Node coordinates, in (rho,theta,zeta).
+
+        Returns
+        -------
+        nodes : ndarray of float, size(num_nodes,3)
+            Node coordinates, in (rho,theta,zeta).
+        spacing : ndarray of float, size(num_nodes,3)
+            Node spacing, in (rho,theta,zeta).
+
+        """
+        nodes = np.atleast_2d(nodes).reshape((-1, 3)).astype(float)
+        # Do not alter nodes given by the user for custom grids.
+        # In particular, do not modulo nodes by 2pi or 2pi/NFP.
+        # This may cause the surface_integrals() function to fail recognizing
+        # surfaces outside the interval [0, 2pi] as duplicates. However, most
+        # surface integral computations are done with LinearGrid anyway.
+        spacing = (  # make weights sum to 4pi^2
+            np.ones_like(nodes) * np.array([1, 2 * np.pi, 2 * np.pi]) / nodes.shape[0]
+        )
+        return nodes, spacing
```

Comment:
Since git struggled with this diff, I just removed the code
```python
self._L = len(np.unique(nodes[:, 0]))
self._M = len(np.unique(nodes[:, 1]))
self._N = len(np.unique(nodes[:, 2]))
```
and assigned them in ``__init__``.